data_scrapers/sonoma.py

```diff
@@ -0,0 +1,299 @@
+#!/usr/bin/env python3
+import requests
+import json
+import re
+import dateutil.parser
+from typing import List, Dict, Union
+from bs4 import BeautifulSoup, element # type: ignore
+from format_error import FormatError # type: ignore
+
+def get_table(header: str, soup: BeautifulSoup) -> element.Tag:
+    """
+    Takes in a header and a BeautifulSoup object and returns the table under
+    that header
+    """
+    header_tag = soup.find(lambda tag: tag.name == 'h3' and header in tag.get_text())
+    tables = header_tag.find_parent().find_all('table')
+    # this lets us get the second cases table
+    return tables[-1]
+
+def get_rows(tag: element.Tag) -> List[element.ResultSet]:
+    """
+    Gets all tr elements in a tag but the first, which is the header
+    """
+    return tag.find_all('tr')[1:]
+
+def get_cells(row: element.ResultSet) -> List[str]:
+    """
+    Gets all th and tr elements within a single tr element
+    """
+    return [el.text for el in row.find_all(['th', 'td'])]
+
+def parse_int(text: str) -> int:
+    text = text.strip()
+    if text == '-':
+        return 0
+    else:
+        return int(text.replace(',', ''))
+
+def generate_update_time(soup: BeautifulSoup) -> str:
+    """
+    Generates a timestamp string (e.g. May 6, 2020 10:00 AM) for when the scraper is run
+    """
+    update_time_text = soup.find('time', {'class': 'updated'}).text.strip()
+    try:
+        date = dateutil.parser.parse(update_time_text)
+    except ValueError:
+        raise ValueError(f'Date is not in ISO 8601'
+                         f'format: "{update_time_text}"')
+    return date.isoformat()
+
+def get_source_meta(soup: BeautifulSoup) -> str:
+    """
+    Finds the 'Definitions' header on the page and gets all of the text in it
+    """
+    definitions_header = soup.find('h3', string='Definitions')
+    definitions_text = definitions_header.find_parent().text
+    return definitions_text.replace('\n', '/').strip()
+
+# apologies for this horror of a output type
+def transform_cases(cases_tag: element.Tag) -> Dict[str, List[Dict[str, Union[str, int, object]]]]:
+    """
+    Takes in a BeautifulSoup tag for the cases table and returns all cases
+    (historic and active), deaths, and recoveries in the form:
+    { 'cases': [], 'deaths': [], 'recovered': [], 'active': [] }
+    Where each list contains dictionaries (representing each day's data)
+    of form (example for cases):
+    { 'date': '', 'cases': -1, 'cumul_cases': -1 }
+    """
+    cases = []
+    cumul_cases = 0
+    deaths = []
+    cumul_deaths = 0
+    # recovered = []
+    # cumul_recovered = 0
+    # active = []
+    # cumul_active = 0
+    rows = get_rows(cases_tag)
+    for row in rows:
+        row_cells = row.find_all(['th', 'td'])
```

Comment:
To be clear, it’s cool if the answer is “tried it, but too much of a pain/not worth the tradeoff.” I just want to make sure it didn’t get missed entirely since I’m concerned about changing columns.

Reply:
Oh, sorry, I missed that suggestion completely. I'll give it a try

Reply:
I started trying this, and I think it's going to be more work than is worth it right now. The shape of the tables has been consistent since I started working on this, although the order of the tables on the page has changed. The function to parse tables is in there, and we can switch over to using it if the table format starts changing

Reply:
OK.

> we can switch over to using it if the table format starts changing

But just to be clear, part of what I was getting at is that it might be hard to notice that the format has started changing. With code like this:

```py
def transform_cases(...):
    ...
    active_cases, new_infected, dead, recoveries = [parse_int(el.text) for el in row_cells[1:]]
    ....
```

If those columns continue to be numeric, nothing about this program will tell you that you’re getting the wrong values out when things were rearranged, new columns were inserted, etc.

Reply:
(This would also help alert you to the fact that the order of tables has changed, since the re-ordered tables would presumably have different headers than you expected, causing errors when you try to get cells by their header.)